src/EditorFeatures/CSharp/AutomaticCompletion/AutomaticLineEnderCommandHandler.cs

```diff
@@ -273,18 +335,917 @@ private static IEnumerable<SyntaxNode> GetOwningNodes(SyntaxNode root, int posit
-                   n is UsingDirectiveSyntax ||
-                   n is ArrowExpressionClauseSyntax;
+            var root = document.GetRequiredSyntaxRootSynchronously(cancellationToken);
+            // Add braces for the selected node
+            if (ShouldAddBraces(selectedNode, caretPosition))
+            {
+                // For these syntax node, braces pair could be easily added by modify the syntax tree
+                if (selectedNode is BaseTypeDeclarationSyntax
+                    or BaseMethodDeclarationSyntax
+                    or LocalFunctionStatementSyntax
+                    or FieldDeclarationSyntax
+                    or EventFieldDeclarationSyntax
+                    or AccessorDeclarationSyntax
+                    or ObjectCreationExpressionSyntax
+                    or WhileStatementSyntax
+                    or ForEachStatementSyntax
+                    or ForStatementSyntax
+                    or LockStatementSyntax
+                    or UsingStatementSyntax
+                    or DoStatementSyntax
+                    or IfStatementSyntax
+                    or ElseClauseSyntax)
+                {
+                    // Add the braces and get the next caretPosition
+                    var (newRoot, nextCaretPosition) = AddBraceToSelectedNode(document, root, selectedNode, cancellationToken);
+                    if (document.Project.Solution.Workspace.TryApplyChanges(document.WithSyntaxRoot(newRoot).Project.Solution))
+                    {
+                        args.TextView.TryMoveCaretToAndEnsureVisible(new SnapshotPoint(args.SubjectBuffer.CurrentSnapshot, nextCaretPosition));
+                    }
+
+                    return;
+                }
+                else
+                {
+                    // For the rest of the syntax node,
+                    // like try statement
+                    // class Bar
+                    // {
+                    //      void Main()
+                    //      {
+                    //          tr$$y
+                    //      }
+                    // }
+                    // In this case, the last close brace of 'void Main()' would be thought as a part of the try statement,
+                    // and the last close brace of 'Bar' would be thought as a part of Main()
+                    // So for these case, just find the missing open brace position and directly insert '()' to the document
+                    // 1. Find the position to insert braces.
+                    var insertionPosition = GetBraceInsertionPosition(selectedNode);
+
+                    // 2. Insert the braces and move caret
+                    InsertBraceAndMoveCaret(args.TextView, document, insertionPosition, cancellationToken);
+                    return;
+                }
+            }
+
+            // Check if we need to remove braces for the node
+            if (ShouldRemoveBraces(selectedNode, caretPosition))
```

Comment:
my general feedback is that examples are good.  just to prime the mind in terms of hwat case this is for.